common/interpreter/wawaka_wasm/WawakaInterpreter.cpp

```diff
@@ -75,43 +78,76 @@ void wasm_printer(const char *msg)
 } /* extern "C" */
 
 // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-void WawakaInterpreter::parse_result_string(
-    int32 result_app,
+void WawakaInterpreter::parse_response_string(
+    int32 response_app,
     std::string& outResult,
-    bool& outStateChanged)
+    bool& outStateChanged,
+    std::map<std::string,std::string>& outDependencies)
 {
-    int32 result_app_beg, result_app_end;
+    // Convert the wasm address for the result string into an
+    // address in the native code
+    int32 response_app_beg, response_app_end;
 
     pe::ThrowIf<pe::RuntimeError>(
-        ! wasm_runtime_get_app_addr_range(wasm_module_inst, result_app, &result_app_beg, &result_app_end),
+        ! wasm_runtime_get_app_addr_range(wasm_module_inst, response_app, &response_app_beg, &response_app_end),
         report_interpreter_error("invalid result pointer", "out of range"));
 
-    char *result = (char*)wasm_runtime_addr_app_to_native(wasm_module_inst, result_app);
-    const char *result_end = result + (result_app_end - result_app);
+    char *result = (char*)wasm_runtime_addr_app_to_native(wasm_module_inst, response_app);
+    const char *result_end = result + (response_app_end - response_app);
 
+    // Not the most performant way to do this, but with no assumptions
+    // about the size of the string, we have to walk the entire string
     for (char *p = result; (*p) != '\0'; p++)
         pe::ThrowIf<pe::RuntimeError>(
             p == result_end,
             report_interpreter_error("invalid result pointer", "unterminated string"));
 
+    SAFE_LOG(PDO_LOG_DEBUG, "response string: %s", result);
+
     // Parse the contract request
     JsonValue parsed(json_parse_string(result));
     pe::ThrowIfNull(
         parsed.value,
         report_interpreter_error("invalid result pointer", "invalid JSON"));
 
-    JSON_Object* parsed_object = json_value_get_object(parsed);
+    // Verify that the response matches the expected schema
+    JsonValue schema(json_parse_string(INVOCATION_RESPONSE_SCHEMA));
+    pe::ThrowIf<pe::RuntimeError>(
+        json_validate(schema.value, parsed.value) != JSONSuccess,
+        "invalid invocation response; does not match required format");
+
+    const JSON_Object* parsed_object = json_value_get_object(parsed);
     pe::ThrowIfNull(
         parsed_object,
         report_interpreter_error("invalid result pointer", "missing result object"));
 
-    outResult.assign(json_object_dotget_string(parsed_object, "Result"));
-    outStateChanged = (json_object_dotget_boolean(parsed_object, "StateChanged") == 1);
+    const char* response_string = json_object_dotget_string(parsed_object, "Response");
+    pe::ThrowIfNull(
```

Comment:
these checks are redundant given the schema validation above, but they do not hurt!